### PR TITLE
Fix Kudu tests to also run on MacOS

### PR DESCRIPTION
**Pull request by grantatspothero**
Resolves https://github.com/trinodb/trino/issues/3415

See the comments for the ugly workaround, but the good news is I'm able to run the tests locally without fixed port mappings.

Going to try stress testing multiple instances of the tests and see what happens

### CODE REVIEW

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 1]
```diff
@@ -15,65 +15,184 @@
...
-                .withCommand("master")
+                .withCommand("/wait_until_port_bound.sh", "master")
+                .withEnv("KUDU_MASTERS", "kudu-master:7051")
+                // Note `rpc_advertised_addresses` does not have a port configured, it will be configured after container startup in `/wait_until_port_bound.sh`
+                .withEnv("MASTER_TEMPLATE_ARGS", format("--fs_wal_dir=/var/lib/kudu/master --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:", masterContainerAlias, KUDU_MASTER_PORT, hostIP))
                 .withNetwork(network)
-                .withNetworkAliases("kudu-master");
+                .withNetworkAliases(masterContainerAlias)
+                // Do not wait, the host port will not be listening until after the container starts up and gets the mapped host port
+                .waitingFor(new NoWaitStrategy());
```
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:88]: Do we need this fix for Master server too ? I think we don't need `rpc_advertised_addresses` for master server

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 2]
```diff
@@ -15,65 +15,184 @@
...
 
         for (int instance = 0; instance < NUMBER_OF_REPLICA; instance++) {
             String instanceName = "kudu-tserver-" + instance;
-            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.10.0")
+            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.15.0")
+                    .withFileSystemBind(tabletTmpDirectories.get(instance), "/testcontainers")
+                    .withCopyFileToContainer(forClasspathResource("wait_until_port_bound.sh", 0777), "/wait_until_port_bound.sh")
                     .withExposedPorts(KUDU_TSERVER_PORT)
-                    .withCommand("tserver")
+                    .withCommand("/wait_until_port_bound.sh", "tserver")
```
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:96]: Instead of using a custom script to bind the address, can we use `ToxiproxyContainer` from TestContainers, specify a proxy and then use the mapped proxy port here.

It could be something like this 
```
        for (int instance = 0; instance < NUMBER_OF_REPLICA; instance++) {
            String instanceName = "kudu-tserver-" + instance;
            ContainerProxy proxy = toxiProxy.getProxy(instanceName, KUDU_TSERVER_PORT);
            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.15.0")
                    .withExposedPorts(KUDU_TSERVER_PORT)
                    .withEnv("KUDU_MASTERS", "kudu-master:" + KUDU_MASTER_PORT)
                    .withEnv("TSERVER_ARGS", format("--fs_wal_dir=/var/lib/kudu/tserver --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:%s", instanceName, KUDU_TSERVER_PORT, hostIP, proxy.getProxyPort()))
                    .withNetwork(network)
                    .withNetworkAliases(instanceName)
                    .waitingFor(new HostPortWaitStrategy())
                    .withCommand("tserver")
                    .dependsOn(master);

            tServersBuilder.add(tableServer);
        }
```

I guess we don't have to bind file in thos case. WDYT ?

Comment (Praveen2112) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:96]: Thanks for fixing this.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 3]
```diff
@@ -15,65 +15,184 @@
...
     {
         Network network = Network.newNetwork();
         ImmutableList.Builder<GenericContainer<?>> tServersBuilder = ImmutableList.builder();
-        this.master = new GenericContainer<>("apache/kudu:1.10.0")
+
+        String hostIP = getHostIPAddress();
+        createHostTmpDirectoriesForVolumeMounts();
+
+        String masterContainerAlias = "kudu-master";
+        this.master = new GenericContainer<>("apache/kudu:1.15.0")
```
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:77]: Version update could be in a different PR.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 2 above)
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:96]: That would definitely be easier than messing with bind mounts! Let me see if I can get it working.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 1 above)
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:88]: Yes you are correct 👍

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 2 above)
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java]: Made things much cleaner, thanks for the suggestion!

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 4]
```diff
@@ -71,9 +102,23 @@ public void close()
         try (Closer closer = Closer.create()) {
             closer.register(master::stop);
             tServers.forEach(tabletServer -> closer.register(tabletServer::stop));
+            closer.register(toxiProxy::stop);
+            closer.register(network::close);
```
Comment (grantatspothero) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:106]: Note networks have to be explicitly closed, since I re-ran these tests a bunch locally I eventually hit up against docker network limits and had to run `docker network prune`. 

This problem also occurs in several other tests:
- https://github.com/trinodb/trino/blob/master/plugin/trino-elasticsearch/src/test/java/io/trino/plugin/elasticsearch/TestElasticsearchBackpressure.java#L41
- https://github.com/trinodb/trino/blob/master/plugin/trino-druid/src/test/java/io/trino/plugin/druid/TestingDruidServer.java#L88

Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:106]: Opened separate PR here: https://github.com/trinodb/trino/pull/10813/files

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 5]
```diff
@@ -18,45 +18,76 @@
...
         ImmutableList.Builder<GenericContainer<?>> tServersBuilder = ImmutableList.builder();
-        this.master = new GenericContainer<>("apache/kudu:1.10.0")
+
+        String hostIP = getHostIPAddress();
+
+        String masterContainerAlias = "kudu-master";
+        this.master = new GenericContainer<>(KUDU_IMAGE)
                 .withExposedPorts(KUDU_MASTER_PORT)
                 .withCommand("master")
+                .withEnv("MASTER_ARGS", format("--fs_wal_dir=/var/lib/kudu/master --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s", masterContainerAlias, KUDU_MASTER_PORT))
```
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:65]: Do we need this change for master server ?

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 6]
```diff
@@ -18,45 +18,76 @@
...
         for (int instance = 0; instance < NUMBER_OF_REPLICA; instance++) {
             String instanceName = "kudu-tserver-" + instance;
-            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.10.0")
+            ToxiproxyContainer.ContainerProxy proxy = toxiProxy.getProxy(instanceName, KUDU_TSERVER_PORT);
+            GenericContainer<?> tableServer = new GenericContainer<>(KUDU_IMAGE)
                     .withExposedPorts(KUDU_TSERVER_PORT)
                     .withCommand("tserver")
-                    .withEnv("KUDU_MASTERS", "kudu-master:" + KUDU_MASTER_PORT)
+                    .withEnv("KUDU_MASTERS", format("%s:%s", masterContainerAlias, KUDU_MASTER_PORT))
+                    .withEnv("TSERVER_ARGS", format("--fs_wal_dir=/var/lib/kudu/tserver --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:%s", instanceName, KUDU_TSERVER_PORT, hostIP, proxy.getProxyPort()))
```
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:81]: I think we could remove `rpc_bind_address` here

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 5 above)
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:65]: When debugging things, having the logs in stderr/available through docker logs was very helpful. 

The rpc bind address config is not strictly necessary since it binds to `0.0.0.0:7051` by default, so I can remove that from the master config.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 6 above)
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:81]: Here I think it is better to be explicit than implicit.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 7]
```diff
@@ -18,45 +18,75 @@
...
+    private static final String TOXIPROXY_IMAGE = "shopify/toxiproxy:2.1.0";
+    private static final String TOXIPROXY_NETWORK_ALIAS = "toxiproxy";
+
+    private final Network network;
+    private final ToxiproxyContainer toxiProxy;
     private final GenericContainer<?> master;
     private final List<GenericContainer<?>> tServers;
 
+    /**
+     * WORKAROUND WARNING:
```
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:48]: Just the comment is fine. 😄 
```suggestion
```

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java [hunk 8]
```diff
@@ -71,9 +101,23 @@ public void close()
...
         catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
+
+    private static String getHostIPAddress()
+    {
+        try {
+            // This is roughly equivalent to setting the KUDU_QUICKSTART_IP defined here: https://kudu.apache.org/docs/quickstart.html#_set_kudu_quickstart_ip
+            // What if there are multiple network interfaces with multiple IPV4 addresses bound? Choose one.
```
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:116]: I don't believe it matters for localhost.

**Review by hashhar — APPROVED** [review]

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 8 above)
Comment (hashhar) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:116]: However if needed we can:

```java
Enumeration<NetworkInterface> networkInterfaceEnumeration = NetworkInterface.getNetworkInterfaces();
while (networkInterfaceEnumeration.hasMoreElements()) {
    for (InterfaceAddress interfaceAddress : networkInterfaceEnumeration.nextElement().getInterfaceAddresses()) {
        if (interfaceAddress.getAddress().isSiteLocalAddress()) {
            return interfaceAddress.getAddress().getHostAddress();
        }
    }
}
```

**Review by hashhar — APPROVED** [review]
LGTM. Thanks a lot for getting this to work.

Please squash commits as `Use ToxiProxy to forward traffic from host to kudu tablets` + some description of why (the javadoc of TestingKuduServer maybe?)

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java (hunk 8 above)
Comment (hashhar) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:115]: Reminder to either remove this comment (if loop-back is just what we need) or address this.

Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java]: Took your suggestion and added a comment explaining why

Comment (hashhar) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java]: Thanks for explaining why loop-back won't work. I didn't even realise at first. 😄

**Review by hashhar — APPROVED** [review]
LGTM.

I'll merge once CI is done.
